README.md
gc_dev
======

CLI tools for TYPO3-developers

Usage:
----------
	./typo3/cli_dispatch.phpsh gc_dev dump-template [<pid>]
		Dumps the template for the given pid. pid defaults to 1.

	./typo3/cli_dispatch.phpsh gc_dev dump-pagetree [<pid>]
		Dumps the pagetree for the given pid. pid defaults to 0.

TODO
----------

* Implement searching / filtering for `dump-template` and `dump-pagetree`.


ChangeLog
2013-03-19   <arvid@gluteus.se>

	* initial code generated with kickstarter


cli/cli.php
<?php

function help($command = FALSE) {
	echo <<<EOT
Usage:

	./typo3/cli_dispatch.phpsh gc_dev dump-template [<pid>]
		Dumps the template for the given pid. pid defaults to 1.

	./typo3/cli_dispatch.phpsh gc_dev dump-pagetree [<pid>]
		Dumps the pagetree for the given pid. pid defaults to 0.
EOT;
}


// Handle arguments
$argv = $_SERVER['argv'];
if (!isset($argv[1])) {
	help();
	exit();
}
$command = $argv[1];
$arguments = array_splice($argv, 2);

// Initialise frontend
if ($command == 'dump-template') {
	// Handle arguments to dump-template
	if (isset($arguments[0])) {
		if (!is_numeric($arguments[0])) {
			help($command);
			exit();
		} else {
			$pid = $arguments[0];
		}
	} else {
		$pid = 1;
	}

	$tsHelper = new tx_gcdev_TSTemplateHelper();
	$template = $tsHelper->getTemplateForPid($pid);
	echo $tsHelper->formatArrayAsTyposcript($template);
} else if ($command == 'dump-pagetree') {
	// Handle arguments to dump-pagetree
	if (isset($arguments[0])) {
		if (!is_numeric($arguments[0])) {
			help($command);
			exit();
		} else {
			$pid = $arguments[0];
		}
	} else {
		$pid = 0;
	}

	// tihi
	$GLOBALS['BE_USER']->user['admin'] = TRUE;

	// Get page node from row
	if ($pid != 0) {
		$pageRow = t3lib_BEfunc::getRecord('pages', $pid);
		$node = t3lib_tree_pagetree_Commands::getNewNode($pageRow);
	} else {
		$nodeData = array('id' => 0);
		$node = t3lib_div::makeInstance('t3lib_tree_pagetree_Node', $nodeData);
		$node->setText('[Root]');
	}

	// Format tree
	$dataProvider = t3lib_div::makeInstance('t3lib_tree_pagetree_DataProvider');
	$formatter = t3lib_div::makeInstance('Tx_GcDev_PageTreeCLIFormatter', $dataProvider);
	$formatter->format($node);
} else {
	help();
	die();
}


?>

Classes/class.Tx_GcDev_TSTemplateHelper.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2010 Arvid Jakobsson <arvid@gluteus.se>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

/* 
 * require_once(PATH_tslib . 'class.tslib_fe.php');
 * require_once(PATH_t3lib . 'class.t3lib_page.php');
 * require_once(PATH_t3lib . 'class.t3lib_userauth.php' );
 * require_once(PATH_tslib . 'class.tslib_feuserauth.php');
 * require_once(PATH_t3lib . 'class.t3lib_timetrack.php');
 */

/**
 * tx_gcsystem_SSFE_TSTemplateHelper
 *
 * description
 *
 * @package 
 * @author  Arvid Jakobsson <arvid@gluteus.se>
 */
class Tx_GcDev_TSTemplateHelper {
	/**
	 * Originally we stored the template in the global value "gc_template". I
	 * think this was because of performance reasons. We do something similar
	 * here by storing the template in a static variable.
	 */
	static private $templatePerPid = array();

	public function getTemplateForPid($pid) {
		// Is template for this pid cached?
		if (isset(self::$templatePerPid[$pid])) {
			$template = self::$templatePerPid[$pid];
		} else {
			/*
			 * initializing tsfe:
			 *  http://lists.typo3.org/pipermail/typo3-english/2007-February/036351.html
			 */
		
			$GLOBALS['TT'] = new t3lib_timeTrack;
			$tsfeClassName = t3lib_div::makeInstanceClassName('tslib_fe');

			$GLOBALS['TSFE'] = new $tsfeClassName($GLOBALS['TYPO3_CONF_VARS'], $pid, '');
			$GLOBALS['TSFE']->showHiddenPage = TRUE;
			$GLOBALS['TSFE']->connectToDB();
			$GLOBALS['TSFE']->initFEuser();
			$GLOBALS['TSFE']->determineId();
			/* $GLOBALS['TSFE']->getCompressedTCarray(); */
			$GLOBALS['TSFE']->initTemplate();
			$GLOBALS['TSFE']->getConfigArray();

			$template = $GLOBALS['TSFE']->tmpl->setup;
			self::$templatePerPid[$pid] = $template;
		}

		return $template;
	}

	public function formatArrayAsTyposcript($template, $ind = 0) {
		foreach ($template as $key => $value) {
			if (is_array($value)) {
				echo str_repeat("\t", $ind) . preg_replace('/.$/', '', $key) . ' {' . PHP_EOL;
				$this->formatArrayAsTyposcript($value, $ind+1);
				echo str_repeat("\t", $ind) . '}' . PHP_EOL;
			} else {
				echo str_repeat("\t", $ind) . $key . ' = ' . $value . PHP_EOL;
			}
		}
	}

}
?>

Classes/class.Tx_GcDev_PageTreeCLIFormatter.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2011 Arvid Jakobsson <arvid@gluteus.se>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

/**
 * Tx_GcDev_PageTreeCLIFormatter
 *
 * Formats a page tree in a format suitable for the CLI, like this:
 *
 * a
 * |-- b
 * |   |-- c
 * |   |-- d
 * |   |   |-- e
 * |   |   `-- f
 * |   `-- g
 * |       `-- h
 * |-- i
 * `-- j
 * `-- k
 *
 * @package
 * @author  Arvid Jakobsson <arvid@gluteus.se>
 */
class Tx_GcDev_PageTreeCLIFormatter {
	protected $dataProvider = NULL;

	public function __construct($dataProvider) {
		$this->dataProvider = $dataProvider;
	}

	public function format($node) {
		$this->formatAux($node, '', '');
	}

	protected function formatAux($node, $ind, $leaf) {
		$label = '[' . $node->getId() . '] ' . (!$node->getText() ? '[null]' : $node->getText());
		echo $leaf . $label  . PHP_EOL;

		$nodeCollection = $this->dataProvider->getNodes($node);

		for ($i = 0, $c = count($nodeCollection); $i < $c; $i++) {
			$node = $nodeCollection[$i];
			$last = ($i == $c - 1);
			$leaf = $ind . ($last ? '`' : '|') . '-- ';

			$ind2 = $last ? $ind . '   ' : $ind . '|  ';
			$this->formatAux($node,  $ind2, $leaf);
		}
	}
}
?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "gc_dev".
#
# Auto generated 19-03-2013 19:22
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Gluteus Development Helpers',
	'description' => '',
	'category' => 'misc',
	'author' => '',
	'author_email' => 'arvid@gluteus.se',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
);

?>